Add more documention on memory management (#4874)
dask/distributed: docs/source/worker.rst

-----------

Each worker sends computations to a thread in a
`concurrent.futures.ThreadPoolExecutor

> <https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor>`_

Author: fixed. I guess that we don't use intersphinx vs. dask/dask due to race conditions in the build process?

the cost of a larger memory footprint. CPython however adds its own memory allocator on
top, which reduces the need for this additional abstraction (with caveats).

Dask.distributed measures the total process memory (also known as RSS) and uses it for
decision-making in several points; namely:

- the spill, pause, and terminate thresholds (see above)
- :meth:`distributed.Client.rebalance` and
  :meth:`distributed.scheduler.Scheduler.rebalance`

In some cases, it is easy to have these heuristics misbehave because there are large
amounts of process memory that the libc did not release to the OS yet - but Dask does
not have the means to tell them apart from a memory leak. This phenomenon is the first
thing you should investigate when you see large amounts of unexplained unmanaged memory
in the dashboard or in the logs.

Manually trim memory
~~~~~~~~~~~~~~~~~~~~
*Linux only*

It is possible to forcefully release allocated but unutilized memory as follows:

.. code-block:: python

    import ctypes

    def trim_memory() -> int:
        libc = ctypes.CDLL("libc.so.6")
        return libc.malloc_trim(0)

    client.run(trim_memory)

> This should be only used as a one-off debugging experiment.

Reviewer: ```suggestion
This should be only used as a one-off debugging experiment. Watch the dashboard while running the above code. If unmanaged worker memory (on the "Bytes stored" plot) decreases significantly after calling ``client.run(trim_memory)``, then move on to the next section. Otherwise, you likely do have a memory leak.

Note that you should only run this ``malloc_trim`` if you are using the default glibc memory allocator. When using a custom allocator, this could cause unexpected behavior including segfaults. (If you don't know what this means, you're probably using the default glibc allocator and are safe to run this.)
```

Automatically trim memory
~~~~~~~~~~~~~~~~~~~~~~~~~
*Linux only*

To aggressively and automatically trim the memory in a production environment, you
should instead set the environment variable ``MALLOC_TRIM_THRESHOLD_`` (note the final

> underscore) to 0 or a low number; see the ``mallopt`` man page for details. Reducing

Reviewer: Also, out of curiosity, did you ever try calling `mallopt` via ctypes? I know it's a little less safe than the env var, but for people using a managed cluster (pangeo, coiled, DIY, etc.) it may be easier to do a `client.run` than setting pre-startup environment variables.

> underscore) to 0 or a low number; see the ``mallopt`` man page for details. Reducing

Author: No, because the mallopt documentation stats out loud that you must invoke it before the very first call to malloc - which in CPython is impossible.

this value will increase the number of syscalls, and as a consequence may degrade

> performance. The variable must be set before starting the dask-worker process.

Reviewer: ```suggestion
performance. **The variable must be set before starting the ``dask-worker`` process.**
```
since people will inevitably miss this and be confused why `client.run(lambda: os.environ.update({"MALLOC_TRIM_THRESHOLD_": "0"}))` doesn't work.

jemalloc
~~~~~~~~
*Linux and MacOS*

Alternatively to the above, you may experiment with the
`jemalloc <http://jemalloc.net>`_ memory allocator, as follows:

.. code-block:: bash

    conda install jemalloc
    LD_PRELOAD=$CONDA_PREFIX/lib/libjemalloc.so dask-worker <...>

jemalloc offers a wealth of configuration settings; please refer to its documentation.

Ignore process memory
~~~~~~~~~~~~~~~~~~~~~

> If all else fails, you may want to stop dask from using the process (RSS) memory in its

Reviewer: ```suggestion
If all else fails, you may want to stop dask from using memory metrics from the OS (RSS) in its
```
just to keep language consistent with other places

      yet (see memory allocators below)
    - Memory fragmentation
    - Memory leaks

unmanaged recent
    Unmanaged memory that has appeared within the last 30 seconds. This is not included
    in the 'unmanaged' memory measure above. Ideally, this memory should be for the most
    part a temporary spike caused by tasks' heap use plus soon-to-be garbage collected
    objects.

    The time it takes for unmanaged memory to transition away from its "recent" state
    can be tweaked through the ``distributed.worker.memory.recent-to-old-time`` key in
    the ``~/.config/dask/distributed.yaml`` file. If your tasks typically run for longer
    than 30 seconds, it's recommended that you increase this setting accordingly.

    By default, :meth:`distributed.Client.rebalance` and
    :meth:`distributed.scheduler.Scheduler.rebalance` ignore unmanaged recent memory.
    This behaviour can also be tweaked using the dask config - see the methods'
    documentation.

spilled
    managed memory that has been spilled to disk. This is not included in the 'managed'
    measure above.

The sum of managed + unmanaged + unmanaged recent is equal by definition to the process
memory.

.. _memtrim:

> Memory not released back to the OS

Reviewer 2: @xhochy I pinged you on another PR recently about explicit malloc trim calls. We discussed this and decided to write this doc page for now. If you have anything to add here, we'd appreciate a short review

Reviewer 3: Thanks for the pointer! Overall this looks good 👍 One could add `mimalloc` as an alternative here, too. We have seen slight performance differences between both in Arrow but I don't think they would have an impact here.

    client.run(trim_memory)

This should be only used as a one-off debugging experiment. Watch the dashboard while
running the above code. If unmanaged worker memory (on the "Bytes stored" plot)
decreases significantly after calling ``client.run(trim_memory)``, then move on to the
next section. Otherwise, you likely do have a memory leak.

Note that you should only run this `malloc_trim`_ if you are using the default glibc
memory allocator. When using a custom allocator such as `jemalloc`_ (see below), this
could cause unexpected behavior including segfaults. (If you don't know what this means,
you're probably using the default glibc allocator and are safe to run this).

Automatically trim memory
~~~~~~~~~~~~~~~~~~~~~~~~~
*Linux workers only*

To aggressively and automatically trim the memory in a production environment, you
should instead set the environment variable ``MALLOC_TRIM_THRESHOLD_`` (note the final
underscore) to 0 or a low number; see the `mallopt`_ man page for details. Reducing
this value will increase the number of syscalls, and as a consequence may degrade
performance. **The variable must be set before starting the ``dask-worker`` process.**

jemalloc

> ~~~~~~~~

Reviewer 3: A further alternative nowadays is `mimalloc`. It works in this case just as `jemalloc` but is available on Windows, too.

      `free`_ yet
    - Unused memory that the user-space libc `free`_ function did not release to the OS
      yet (see memory allocators below)
    - Memory fragmentation
    - Memory leaks

unmanaged recent
    Unmanaged memory that has appeared within the last 30 seconds. This is not included
    in the 'unmanaged' memory measure above. Ideally, this memory should be for the most
    part a temporary spike caused by tasks' heap use plus soon-to-be garbage collected
    objects.

    The time it takes for unmanaged memory to transition away from its "recent" state
    can be tweaked through the ``distributed.worker.memory.recent-to-old-time`` key in
    the ``~/.config/dask/distributed.yaml`` file. If your tasks typically run for longer
    than 30 seconds, it's recommended that you increase this setting accordingly.

    By default, :meth:`distributed.Client.rebalance` and
    :meth:`distributed.scheduler.Scheduler.rebalance` ignore unmanaged recent memory.
    This behaviour can also be tweaked using the dask config - see the methods'
    documentation.

spilled
    managed memory that has been spilled to disk. This is not included in the 'managed'
    measure above.

> The sum of managed + unmanaged + unmanaged recent is equal definition to the process

Reviewer: ```suggestion
The sum of managed + unmanaged + unmanaged recent is equal by definition to the process
```

Dask — it's actually normal behavior for all processes on Linux and MacOS, and is a
consequence of how the low-level memory allocator works (see below for details).

Because Dask makes decisions (spill-to-disk, pause, terminate,
:meth:`~distributed.Client.rebalance`) based on the worker's memory usage as reported by
the OS, and is unaware of how much of this memory is actually in use versus empty and
"hoarded", it can overestimate — sometimes significantly — how much memory the process
is using and think the worker is running out of memory when in fact it isn't.

More in detail:

Both the Linux and MacOS memory allocators try to avoid performing a `brk`_ kernel call
every time the application calls `free`_ by implementing a user-space memory management
system. Upon `free`_, memory can remain allocated in user space and potentially reusable
by the next `malloc`_ - which in turn won't require a kernel call either. This is
generally very desirable for C/C++ applications which have no memory allocator of their
own, as it can drastically boost performance at the cost of a larger memory footprint.
CPython however adds its own memory allocator on top, which reduces the need for this

> additional abstraction (with caveats).

Author: I don't think .. note:: is appropriate as it's a paragraph that deserves _less_ attention, whereas .. note:: signals it deserves more. I've tweaked it slightly.

> Manually trim memory

Author: done
